php/class-customize-snapshot-manager.php

```diff
@@ -0,0 +1,559 @@
+
+		// Enqueue scripts.
+		wp_enqueue_script( $this->plugin->script_handles[ $handle ] );
+
+		// Set the sanpshot theme.
+		$snapshot_theme = null;
+		if ( isset( $this->snapshot->post()->ID ) ) {
+			$snapshot_theme = get_post_meta( $this->snapshot->post()->ID, '_snapshot_theme', true );
+		}
+
+		// Script data array.
+		$exports = array(
+			'nonce' => wp_create_nonce( self::AJAX_ACTION ),
+			'action' => self::AJAX_ACTION,
+			'uuid' => $this->snapshot->uuid(),
+			'is_preview' => $this->snapshot->is_preview(),
+			'snapshot_theme' => $snapshot_theme,
+			'scope' => ( isset( $_GET['scope'] ) ? $_GET['scope'] : 'dirty' ),
+			'i18n' => array(
+				'saveButton' => __( 'Save', 'customize-widgets-plus' ),
+				'cancelButton' => __( 'Cancel', 'customize-widgets-plus' ),
+				'shareButton' => __( 'Share URL to preview', 'customize-widgets-plus' ),
+				'updateMsg' => __( 'Clicking "Save" will update the current snapshot.', 'customize-widgets-plus' ),
+				'errorMsg' => __( 'The snapshot could not be saved.', 'customize-widgets-plus' ),
+				'previewTitle' => __( 'Preview URL', 'customize-widgets-plus' ),
+				'formTitle' => ( $this->snapshot->is_preview() ?
+					__( 'Update Snapshot', 'customize-widgets-plus' ) :
+					__( 'Snapshot Scope', 'customize-widgets-plus' )
+				),
+				'dirtyLabel' => __( 'Diff Snapshot (preview dirty settings)', 'customize-widgets-plus' ),
+				'fullLabel' => __( 'Full Snapshot (preview all settings)', 'customize-widgets-plus' ),
+			),
+		);
+
+		// Export data to JS.
+		wp_scripts()->add_data(
+			$this->plugin->script_handles[ $handle ],
+			'data',
+			sprintf( 'var _customizeWidgetsPlusCustomizeSnapshot = %s;', wp_json_encode( $exports ) )
+		);
+	}
+
+	/**
+	 * Get the Customize_Snapshot instance.
+	 *
+	 * @return Customize_Snapshot
+	 */
+	public function snapshot() {
+		return $this->snapshot;
+	}
+
+	/**
+	 * Save a snapshot.
+	 *
+	 * @param \WP_Customize_Manager $manager WP_Customize_Manager instance.
+	 * @param string $status The post status.
+	 * @return null|\WP_Error
+	 */
+	public function save( \WP_Customize_Manager $manager, $status = 'draft' ) {
+		$new_setting_ids = array_diff( array_keys( $this->post_data ), array_keys( $manager->settings() ) );
+		$manager->add_dynamic_settings( $new_setting_ids );
+
+		if ( ! empty( $this->contextual_post_data ) ) {
+			foreach ( $this->contextual_post_data as $setting_id => $contextual ) {
+				$this->snapshot->set_contextual( $setting_id, $contextual['value'], $contextual['dirty'] );
+			}
+		}
+
+		foreach ( $manager->settings() as $setting ) {
+			if ( $this->can_preview( $setting, $this->post_data ) ) {
+				$post_data = $this->post_data[ $setting->id ];
+				$this->snapshot->set( $setting, $post_data['value'], $post_data['dirty'] );
+			}
+		}
+
+		$r = $this->snapshot->save( $status );
+		if ( is_wp_error( $r ) ) {
+			status_header( 500 );
+			wp_send_json_error( $r->get_error_message() );
```

Comment:
@valendesigns instead of making `\CustomizeWidgetsPlus\Customize_Snapshot_Manager::save()` assume that it is being called in an Ajax request context, let's just have it return this `$r` right here. Then `\CustomizeWidgetsPlus\Customize_Snapshot_Manager::save_snapshot()` can handle the `WP_Error` and amend the `customize_save_response` with the error message for the Customizer JS to receive via `wp.customize.bind( 'saved', function ( response ) { ... } )`